File: bigger-is-greater/bigger-is-greater.go
package biggerisgreater

func biggerIsGreater(word string) string {
	chars := []rune(word)

	pivotIdx := -1
	for i := 1; i < len(chars); i++ {
		if chars[i] > chars[i-1] {
			pivotIdx = i - 1
		}
	}

	if pivotIdx == -1 {
		return "no answer"
	}

	pivot := chars[pivotIdx]
	pivotIdxNext := -1
	for i := pivotIdx; i < len(chars); i++ {
		if chars[i] > pivot {
			pivotIdxNext = i
		}
	}

	chars[pivotIdx] = chars[pivotIdxNext]
	chars[pivotIdxNext] = pivot

	result := ""
	for i := 0; i <= pivotIdx; i++ {
		result += string(chars[i])
	}
	for i := len(chars) - 1; i > pivotIdx; i-- {
		result += string(chars[i])
	}

	if result == word {
		return "no answer"
	} else {
		return result
	}
}

File: bigger-is-greater/bigger-is-greater.js
const biggerIsGreater = word => {
  const chars = [ ...word ];

  const pivotIdx = chars
    .slice(1)
    .reduce((acc, ch, i) => ch > chars[i] ? i : acc, -1);

  if (pivotIdx === -1) {
    return 'no answer';
  }

  const pivot = chars[pivotIdx];

  const pivotIdxNext = chars
    .slice(pivotIdx)
    .reduce((acc, ch, i) => ch > pivot ? i + pivotIdx : acc, -1);

  chars[pivotIdx] = chars[pivotIdxNext];
  chars[pivotIdxNext] = pivot;

  const result = [
    ...chars.slice(0, pivotIdx + 1),
    ...chars.slice(pivotIdx + 1).reverse()
  ].join('');

  return result === word
    ? 'no answer'
    : result;
};
